### PR TITLE
Refactor/payment and yapily domain rework

### PR DESCRIPTION
**Pull request by hwarrior**
Removed Biller validation from controller.
Added YapilyService (as domain service) which returns the auth. url.
Renamed IOpenBankingService to IYapilyService - we won't have other providers for now.
Changed PaymentRequest to contain the SmartRequest itself and added a repository for fetching PaymentRequests.
Renamed Amount to AmountToPay in the PaymentRequest to differentiate from the total outstanding amount on the SmartRequest.
Renamed Amount in SmartRequest to OutstandingAmount.
Added an exception for validation errors in the domain.
Added validation logic in PaymentRequest builder for checking if the requested amount to paid is valid.
Added domain model to represent a PaymentAuthorisation.
Removed senderAccountId from the Yapily service request.
Mapped actual values from the domain PaymentRequest to the Yapily service request.

Updated unit and integration tests.

### CODE REVIEW

File: src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs [hunk 1]
```diff
@@ -10,47 +10,43 @@ namespace Ordo.OpenBankingService.Infrastructure.Repositories.Providers.Yapily
...
-
-
-        // TODO move the logic of checking for existence out from infrastructure into the domain(and probably have just one create method)
         // TODO handle error cases when new mapping is not successful
         // TODO handle case when Yapily user already exist but there is no mapping on our end
-        public async Task<IUser> CreateYapilyUser(IUser user)
+        public async Task<IYapilyUser> CreateYapilyUser(Guid platformUuid)
         {
-            if (user == null)
+            if (platformUuid == null)
```
Comment (tombiddulph) [inline src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs]: Guid is a value type (struct) so can't be null
To check that it has an acceptable value you can use

`platformUuid == Guid.Empty` 

or 
`platformUuid == default` (not 100% sure on this variation)

File: src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs [hunk 2]
```diff
@@ -10,47 +10,43 @@ namespace Ordo.OpenBankingService.Infrastructure.Repositories.Providers.Yapily
...
             await _userContext.SaveChangesAsync();
 
             return entityResult.Entity;
         }
 
-        public async Task<IUser> UserExists(IUser user)
+        public async Task<IYapilyUser> GetYapilyUser(Guid platformUuid)
         {
-            if (user == null)
+            if (platformUuid == null)
```
Comment (tombiddulph) [inline src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs]: Same as the constructor

File: src/ApiTest/PaymentControllerTests.cs [hunk 3]
```diff
@@ -37,118 +36,89 @@ public class PaymentControllerTests
...
-        [InlineData(BILLER_ID, ACCOUNT_NUMBER, null, DISPLAY_NAME)] // missing sort code
-        [InlineData(BILLER_ID, null, SORT_CODE, null)] // missing account number
-        [InlineData(null, ACCOUNT_NUMBER, SORT_CODE, null)] // missing biller id
-        public async Task Post_InitiatePayment_SmartRequestPlatformError_ExceptionThrown(
-            string billerId, string accountNumber, string sortCode, string displayName)
-        {
-            //arrange
-            var biller = Biller.Build(billerId, accountNumber, sortCode, displayName);
-            ISmartRequest sr = new SmartRequest(SMART_REQUEST_ID, REFERENCE, biller);
+        //[Theory]
```
Comment (tombiddulph) [inline src/ApiTest/PaymentControllerTests.cs]: Was this meant to be deleted

File: src/Domain/Errors/DomainErrorMessages.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Ordo.OpenBankingService.Domain.Errors
+{
+    public static class DomainErrorMessages
+    {
+        public const string AMOUNT_NOT_PAYABLE = "The requested amount to pay is not payable for the specified Smart Request.";
```
Comment (tombiddulph) [inline src/Domain/Errors/DomainErrorMessages.cs]: Should we move this to config?

File: src/Domain/Payment/Model/PaymentRequest.cs [hunk 5]
```diff
@@ -1,43 +1,44 @@
...
             if (string.IsNullOrEmpty(institutionId))
                 throw new ArgumentNullException(nameof(institutionId));
 
+            ValidatePaymentAmount(smartRequest, amountToPay);
+
             return new PaymentRequest
             {
-                Amount = amount,
-                SmartRequestId = smartRequestId,
+                AmountToPay = amountToPay,
```
Comment (tombiddulph) [inline src/Domain/Payment/Model/PaymentRequest.cs]: This is a prime opportunity to use some kind of mapping library

File: src/Domain/Provider/YapilyService.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        private readonly IYapilyUserRepository _userRepository;
+
+        public YapilyService(IPaymentRequestRepository paymentRequestRepo, IPaymentRepository paymentRepository, IYapilyUserRepository userRepository)
+        {
+            _paymentRequestRepo = paymentRequestRepo ?? throw new ArgumentNullException(nameof(paymentRequestRepo));
+            _paymentRepository = paymentRepository ?? throw new ArgumentNullException(nameof(paymentRepository));
+            _userRepository = userRepository ?? throw new ArgumentNullException(nameof(userRepository));
+        }
+
+        public async Task<IPaymentAuthorisation> InitiatePayment(Guid platformUuid, string smartRequestId, decimal amount, string currency, string institutionId)
```
Comment (tombiddulph) [inline src/Domain/Provider/YapilyService.cs]: Should we be validating parameters?

File: src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs [hunk 7]
```diff
@@ -13,14 +13,14 @@ private YapilyInitiatePaymentRequest()
...
-        public YapilyInitiatePaymentRequest(IPaymentRequest paymentRequest, IUser user, string callbackUrl)
+        public YapilyInitiatePaymentRequest(IPaymentRequest paymentRequest, IYapilyUser user, string callbackUrl)
         {
             if (paymentRequest == null)
             {
                 throw new ArgumentNullException(nameof(paymentRequest));
             }
 
-            if (user == null || !(user is User u))
+            if (user == null || !(user is YapilyUser yapilyUser))
```
Comment (tombiddulph) [inline src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs]: I don't think it's correct to throw an ArgumentNullException for the right handside of the expression, I think I wrote that code :D

File: src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs [hunk 8]
```diff
@@ -30,7 +30,7 @@ public YapilyInitiatePaymentRequest(IPaymentRequest paymentRequest, IUser user,
                 throw new ArgumentException(nameof(callbackUrl));
             }
 
-            UserUuid = u.YapilyUuid.ToString();
+            UserUuid = yapilyUser.YapilyUuid.ToString();
```
Comment (tombiddulph) [inline src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs]: Again, a prime opportunity for a mapping library

File: src/Infrastructure/Repositories/PaymentRequestRepository.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    internal class PaymentRequestRepository : IPaymentRequestRepository
+    {
+        private readonly ISmartRequestRepository _smartRequestRepository;
+
+        public PaymentRequestRepository(ISmartRequestRepository smartRequestRepository)
+        {
+            _smartRequestRepository = smartRequestRepository ?? throw new ArgumentNullException(nameof(smartRequestRepository));
+        }
+
+        public async Task<IPaymentRequest> GetPaymentRequest(string smartRequestId, decimal amountToPay, string currency, string institutionId)
```
Comment (tombiddulph) [inline src/Infrastructure/Repositories/PaymentRequestRepository.cs]: Parameter validation?

File: src/Api/Controllers/PaymentController.cs [hunk 10]
```diff
@@ -39,43 +30,28 @@ public PaymentController(ISmartRequestBuilder smartRequestBuilder, IOpenBankingP
         /// </remarks>
         /// <returns></returns>
         [HttpPost]
-        public async Task<ActionResult<DummyResponse>> InitiatePayment(
-            [FromBody] InitiatePaymentRequest request)
+        public async Task<ActionResult<DummyResponse>> InitiatePayment([FromBody] InitiatePaymentRequest request)
```
[the remark this answers is not in the document]
Comment (hwarrior) [reply src/Api/Controllers/PaymentController.cs]: Thanks, I've also added a todo to do it soon after:)
I can't see the model though in Api?

File: src/Domain/Errors/BaseException.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Ordo.OpenBankingService.Domain.Errors
+{
+    public abstract class BaseException : Exception
+    {
+        protected BaseException(string errorMessage) : base(errorMessage)
```
[the remark this answers is not in the document]
Comment (hwarrior) [reply src/Domain/Errors/BaseException.cs]: Good point, deleted.
Started doing some work on cleaning up our exceptions a bit but stopped.

File: src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs (hunk 1 above)
Comment (hwarrior) [reply src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs]: Nice catch! I've changed it at a couple of other places as well.

File: src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs (hunk 2 above)
Comment (hwarrior) [reply src/Infrastructure/Repositories/Providers/Yapily/YapilyUserRepository.cs]: Fixed

File: src/ApiTest/PaymentControllerTests.cs (hunk 3 above)
Comment (hwarrior) [reply src/ApiTest/PaymentControllerTests.cs]: Deleted.

File: src/Domain/Errors/DomainErrorMessages.cs (hunk 4 above)
Comment (hwarrior) [reply src/Domain/Errors/DomainErrorMessages.cs]: No, error messages shouldn't change all that much(or at all). Unless the intention is to display these directly on the UI - however thats not the case here.

File: src/Domain/Payment/Model/PaymentRequest.cs (hunk 5 above)
Comment (hwarrior) [reply src/Domain/Payment/Model/PaymentRequest.cs]: Agree - however first we had to clean up our domain. Mapping comes soon.

File: src/Domain/Provider/YapilyService.cs (hunk 6 above)
Comment (hwarrior) [reply src/Domain/Provider/YapilyService.cs]: Yes!
The idea is to validate the input parameters in the Api layer once and then validate them again in the Domain when we build our domain objects - ensuring that once a domain object is created it is also valid.
A lot of this is missing and we just relying on the Api at the moment to supply the right parameters. We also don't have validation rules for some things.
I've opened https://github.com/smartrequestco/open-banking-service/issues/34 to track this.

File: src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs (hunk 7 above)
Comment (hwarrior) [reply src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs]: Good catch:) I've removed that cast and check as at this point we only have IYapilyUser.

File: src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs (hunk 8 above)
Comment (hwarrior) [reply src/Infrastructure/Dto/Yapily/YapilyInitiatePaymentRequest.cs]: noted once more:)

File: src/Infrastructure/Repositories/PaymentRequestRepository.cs (hunk 9 above)
Comment (hwarrior) [reply src/Infrastructure/Repositories/PaymentRequestRepository.cs]: Same answer as above.

**Pull request merged**